### ISSUE TITLE
ranger becomes unresponsive when openning image with drag-on-drop

### ISSUE THREAD

**Issue by aranmohyeddin** [ranger/ranger#1890]
#### Runtime Environment
- Operating system and version: Archlinux
- Terminal emulator and version: kitty 0.16
- ranger version: ranger 1.9.3
- Python version: 3.8.2 (default, Feb 26 2020, 22:21:03) [GCC 9.2.1 20200130]
- Locale: en_US.UTF-8

#### Steps to reproduce
1. open_width [dragon-drag-and-drop](https://github.com/schne324/dragon-drop)
2. ctrl+c to close dragon-drop 
3. ranger is not responding anymore

**Comment by toonn** [comment]
Does this happen if you close dragon-drop a different way?

**Comment by aranmohyeddin** [comment]
@toonn yes, if I kill it with a signal it also happens. Also, I don't have anything to open images with by default. and if I press `l` to open an image the same thing happens.

**Comment by toonn** [comment]
Hmm, what options are listed when you press `r` while one of these images is highlighted?

**Comment by aranmohyeddin** [comment]
Actually it becomes buggy when I press r! The cursor jumps in the middle of the page and after a few seconds I get the following text in the bottom and it disappears after a few seconds.
```
1 | ${VISUAL:-$EDITOR} -- "$@"
2 | "$PAGER" -- "$@"
:open_with
```

**Comment by toonn** [comment]
Did this happen before you setup the dragon-drop integration?

**Comment by aranmohyeddin** [comment]
You mean [this](https://github.com/ranger/ranger/wiki/Drag-and-Drop)? I haven't done this. I was not aware it existed. I just added the rifle way but I'm still having the issue.

**Comment by toonn** [comment]
And the issue persists when you run `ranger --clean`?

**Comment by aranmohyeddin** [comment]
No. `ranger --clean` works fine. Not offense, but I think if you had asked all three questions in the first place we would have been able to proceed faster, if you would like we can also chat on IRC.

**Comment by toonn** [comment]
If `--clean` works fine it's an issue with your configuration. Try to narrow it down to as few changes as possible. Feel free to ask in the IRC as well. I would've asked everything in one comment if I'd thought of everything at the same time...

**Comment by aranmohyeddin** [comment]
The problem is that I have these three lines in my rc.lua

**Comment by toonn** [comment]
So it has nothing to do with dragon?

**Comment by aranmohyeddin** [comment]
@toonn apparently no...

**Comment by aranmohyeddin** [comment]
After some troubleshooting with toonn on IRC, here is the situation:
 * My scope.sh and rifle.conf are up to date.
 * Using ueberzug in kitty does not have this issue. (but it's not a solution because ueberzug doesn't respect kitty windows or tabs)
 * Using terminology method works just fine in terminology terminal.

The most confusing thing is:
If instead of putting 
```
set preview_images true
set preview_images_method kitty
```
In rc.conf,
I run `ranger --clean` and then manually do those sets with `:set preview_images true` and `:set preview_images_method kitty` it works perfectly fine!
But if I add those two lines to rc.conf and then open ranger I get the bug.

@markdawn do you have any thoughts?

**Comment by aranmohyeddin** [comment]
@toonn I believe you should add the `bug` label.

**Comment by wis** [comment]
you should narrow down your description of the issue, 
when using the kitty preview method and I press the first character of a sequence and ranger draws the next characters of the sequence I can press, it hangs.
$ `ranger --version`
```
ranger version: ranger 1.9.3
Python version: 3.8.2 (default, Feb 26 2020, 22:21:03) [GCC 9.2.1 20200130]
Locale: en_US.UTF-8
```
$ `kitty -v`
`kitty 0.17.2 created by Kovid Goyal`
$ `uname -a` 
`Linux archsystem 5.5.13-arch2-1 #1 SMP PREEMPT Mon, 30 Mar 2020 20:42:41 +0000 x86_64 GNU/Linux`

**Comment by toonn** [comment]
Hmm, are you sure it's a regression in ranger and not kitty? Nothing changed to the kitty preview method specifically.

**Comment by wis** [comment]
it is conceivable that the issue could be drawing ncurses TUI on top of the rendered image by the terminal emulator, kitty, or vice versa. I'll open an issue at [kovidgoyal/kitty](https://github.com/kovidgoyal/kitty) for this, if there isn't any.

**Comment by toonn** [comment]
The delays may be because the execution of the `<bg>` binding is synchronous.
Let's not open a kitty issue unless we have some evidence pointing to a kitty regression.

**Comment by wis** [comment]
> The delays may be because the execution of the <bg> binding is synchronous.
No, it's not it, because I commented out all bindings that has a `<bg>` for that command/key (`y`),  
after couple tests I found that there's a delay only when there's something previewed with kitty preview, after pressing the key the previewed image disappears immediately but after that it takes ranger a while to show the menu.
> Let's not open a kitty issue unless we have some evidence pointing to a kitty regression.
I agree.

**Comment by aranmohyeddin** [comment]
Another issue that I found is:
1. set preview method to `ueberzug` in rc.conf
2. open kitty and navigate to an image
3. run `:set preview_images_method kitty`
> kitty won't respond anymore
but,
1. set preview method to `kitty` in rc.conf
2. open kitty and navigate to an image
3. run `:set preview_images_method ueberzug`
> kitty still responds

**Comment by jevy** [comment]
> I run `ranger --clean` and then manually do those sets with `:set preview_images true` and `:set preview_images_method kitty` it works perfectly fine!

I have the same issue with kitty. Though when I do the above commands it does not work fine. 

Example here: https://asciinema.org/a/KO2CQX6Isug4OQQHgocE0VtxM

Note: image previews work fine, but don't show up in Asciinema. You will see how things get weird when I press `r`

**Comment by fabh2o** [comment]
# In kitty terminal:

- Method `kitty`: If the image to display is big (tested with 1080x1350 ), the image is shown at 100% causing it to go out of the terminal windows ![ranger-kitty-size](https://user-images.githubusercontent.com/13369709/104215116-9d5dbd00-5438-11eb-81ba-76304918a148.png)


- Method `ueberzug`: images are straight up not displayed, maybe because they are scaled wrongly and so the `x` and `y` "starting" point are already out of view

# In alacritty:

-  Method `ueberzug`: workig as expected, shows the right size
![ranger-alacritty-ueb-right-size](https://user-images.githubusercontent.com/13369709/104218161-a0f34300-543c-11eb-972e-d435cbbc7978.png)



Problem has been seen on Manjaro (xfce), but testing in a clean system (EndeavourOs virtualbox) show the same results
```bash
fabh2o@endOs-virtualbox ~> pip freeze | grep -P '(ueb)|(ranger)'
    ranger-fm==1.9.3
    ueberzug==18.1.9
fabh2o@endOs-virtualbox ~> pacman -Q kitty alacritty
    kitty 0.19.3-1
    alacritty 0.6.0-1
fabh2o@endOs-virtualbox ~> cat .config/ranger/rc.conf 
    set preview_images true
    set preview_images_method kitty
    set draw_borders true
DesktopEnv: xfce
```

**Comment by toonn** [comment]
I don't understand how the kitty method could be drawing outside the terminal window.

@kovidgoyal, it's impossible for kitty's image protocol to cause drawing of an image outside of the bounds of the terminal window, no?

**Comment by fabh2o** [comment]
> I don't understand how the kitty method could be drawing outside the terminal window.
> 

I meant that a portion of the image goes out of the visible space inside the terminal window, not that is rendered outside.
I didn't choose my words very well, sorry

Screenshot: https://i.imgur.com/e91ZKnf.png

**Comment by toonn** [comment]
Not sure I understand. What was the earlier screenshot where the image goes far beyond ranger's borders then?

**Comment by fabh2o** [comment]
I hope the video is clear, the image never gets outside of the terminal window, but goes outside of the space that ranger predispose for the image preview.

https://user-images.githubusercontent.com/13369709/104356261-05310800-550c-11eb-8ced-7a5478ed7d0b.mp4

**Comment by toonn** [comment]
Thanks for the video, that's clearer. So when the preview is displayed, ranger's UI disappears until you press another key?

**Comment by fabh2o** [comment]
It doesnt desappear, it get shifted to make space to the image, see the first [screenshot](https://user-images.githubusercontent.com/13369709/104215116-9d5dbd00-5438-11eb-81ba-76304918a148.png). In this case the terminal windows is bigger and we see that the ranger interface gets "shifted".
I think that the top left of the image is in the right place, but since the image is not scaled down it gets displayed in full size. Since there is no vertical space to display the whole height of the image, the terminal ""scrolls"" until the bottom of the image is visible, which cause the ranger interface to go up (you can see the line that displays file attribute and the date).
In the video you can see that if the terminal has enough lines we can still see the bottom line of ranger, if the terminal is smaller that line is not visible (shifted too high)

https://user-images.githubusercontent.com/13369709/104361667-0f0a3980-5513-11eb-81d5-92795da1182c.mp4

**Comment by toonn** [comment]
And you have the python library `pillow` installed, right? It's what the kitty preview method uses to resize images.

**Comment by fabh2o** [comment]
Yes, `pip freeze` shows `Pillow==8.1.0`

**Comment by fabh2o** [comment]
I tried `kitty +kitten icat --place 20x20@0x0 image.jpg` (i think ranger use something like this) and the image is in no way scaled down, so the problem is `kitty` rather than ranger itself. But this doesnt explain why the `ueberzug` method doesnt work in kitty (look at my [first post](https://github.com/ranger/ranger/issues/1890#issuecomment-758115167))

**Comment by toonn** [comment]
Kitty doesn't do the scaling, ranger scales and then sends to kitty for display. I don't know why ueberzug wouldn't work with kitty.

**Comment by Sheco** [comment]
I'm facing the same issue.

**Comment by LRitzdorf** [comment]
I've got a potentially easier (?) reproduction for this freeze, if that would help.
```sh
$ kitty ranger --clean
:set preview_images_method kitty
:set preview_images true
# select an image to trigger preview
:shell echo hello
# Ranger freezes
```
If `preview_images` is not enabled, the freeze does not occur (as we expect, at this point).

Please let me know if there's any further troubleshooting I can do!

Also, for ease cross-referencing: kovidgoyal/kitty#3864 (closed; don't necrobump this)

**Comment by arobase-che** [comment]
@LRitzdorf: Please, verify that your are using up to date ranger. Arch Linux uses release tag but the this repository doesn't publish any new version for a while.

The bug you are talking about should have been fixed with #1758.

**Comment by LRitzdorf** [comment]
...yikes, that last release is *very* old. I see some pretty recent commits; any idea why there hasn't been a release since 2019?
